hasktorch/src/Torch/NN.hs

```diff
@@ -41,73 +43,114 @@ nextParameter = do
     [] -> error "Not enough parameters supplied to replaceParameters"
     (p : t) -> do put t; return p
 
-class HasForward f a b | f a -> b where
+class HasForward f a where
+  type B f a :: Type
   forward :: f -> a -> b
-  default forward ::
-    ( Generic f,
-      Generic a,
-      Generic b,
-      GHasForward (Rep f) (Rep a) (Rep b)
-    ) =>
-    f ->
-    a ->
-    b
-  forward f a = to $ gForward (from f) (from a)
-  forwardStoch :: f -> a -> IO b
-  default forwardStoch ::
-    ( Generic f,
-      Generic a,
-      Generic b,
-      GHasForward (Rep f) (Rep a) (Rep b)
-    ) =>
-    f ->
-    a ->
-    IO b
-  forwardStoch f a = to <$> gForwardStoch (from f) (from a)
-
-class GHasForward (f :: Type -> Type) (a :: Type -> Type) (b :: Type -> Type) | f a -> b where
-  gForward :: forall c c' c''. f c -> a c' -> b c''
-  gForwardStoch :: forall c c' c''. f c -> a c' -> IO (b c)
-
-instance GHasForward U1 U1 U1 where
-  gForward U1 U1 = U1
-  gForwardStoch U1 U1 = return U1
-
-instance
-  ( GHasForward f a b,
-    GHasForward g a' b',
-    b'' ~ (b :+: b')
-  ) =>
-  GHasForward (f :+: g) (a :+: a') b''
-  where
-  gForward (L1 f) (L1 a) = L1 $ gForward f a
-  gForward (R1 g) (R1 a') = R1 $ gForward g a'
-  gForwardStoch (L1 f) (L1 a) = L1 <$> gForwardStoch f a
-  gForwardStoch (R1 g) (R1 a') = R1 <$> gForwardStoch g a'
-
-instance
-  ( GHasForward f a b,
-    GHasForward g a' b',
-    b'' ~ (b :*: b')
-  ) =>
-  GHasForward (f :*: g) (a :*: a') b''
-  where
-  gForward (f :*: g) (a :*: a') = gForward f a :*: gForward g a'
-  gForwardStoch (f :*: g) (a :*: a') = liftA2 ((:*:)) (gForwardStoch f a) (gForwardStoch g a')
 
-instance
-  (HasForward f a b) =>
-  GHasForward (K1 i f) (K1 i a) (K1 i b)
-  where
-  gForward (K1 f) (K1 a) = K1 $ forward f a
-  gForwardStoch (K1 f) (K1 a) = K1 <$> forwardStoch f a
+data ModelRandomness = Deterministic | Stochastic
 
-instance
-  (GHasForward f a b) =>
-  GHasForward (M1 i t f) (M1 i t' a) (M1 i t' b)
-  where
-  gForward (M1 f) (M1 a) = M1 $ gForward f a
-  gForwardStoch (M1 f) (M1 a) = M1 <$> gForwardStoch f a
+type family ModelRandomnessR (out :: Type) :: ModelRandomness where
+  ModelRandomnessR ((->) g (_, g)) = 'Stochastic
```

Comment:
Until the dust settles and we've figured out what random generator state to use, you may want to use the concrete `G` data type here, too. Otherwise, this will match also on, say, `Int -> (_, Int)`.

To make this even better, we could write another type family that deconstructs the passed in type constructor by constructor and checks that `G` doesn't appear any other way than the above in it. If it does, we return a `TypeError`, and GHC will tell the user they this ain't good 😊

Reply:
Added a check in 04b4175 (it can probably be improved!)